src/fetch.ts

```diff
@@ -27,24 +27,35 @@ export interface $Fetch {
   raw<T = any, R extends ResponseType = 'json'>(request: FetchRequest, opts?: FetchOptions<R>): Promise<FetchResponse<MappedType<R, T>>>
 }
 
-export function setHeader (options: FetchOptions, _key: string, value: string) {
-  const key = _key.toLowerCase()
-  options.headers = options.headers || {}
-  if ('set' in options.headers) {
-    ;(options.headers as Headers).set(key, value)
-  } else if (Array.isArray(options.headers)) {
-    const existingHeader = options.headers.find(([header]) => header.toLowerCase() === key)
-    if (existingHeader) {
-      existingHeader[1] = value
-    } else {
-      options.headers.push([key, value])
-    }
+export function normalizeHeaders (_opts: FetchOptions): Pick<Headers, 'get' | 'set' | 'has'> {
+  const opts = (_opts as FetchOptions & { _headers: Pick<Headers, 'get' | 'set' | 'has'> })
+  if (opts._headers) { return opts._headers }
+
+  opts.headers = opts.headers || {}
+  if ('set' in opts.headers) { return opts.headers as Pick<Headers, 'get' | 'set' | 'has'> }
+
+  let headers: Record<string, string>
+  if (Array.isArray(opts.headers)) {
+    headers = Object.fromEntries(opts.headers.map(([key, value]) => [key.toLowerCase(), value]))
   } else {
-    const existingHeader = Object.keys(options.headers).find(header => header.toLowerCase() === key)
-    options.headers[existingHeader || key] = value
+    headers = Object.fromEntries(Object.entries(opts.headers).map(([key, value]) => [key.toLowerCase(), value]))
   }
+
+  opts.headers = headers
+  opts._headers = {
```

Comment:
Mutating user options is probably not the best idea as it can affect next calls if they use shared reference and implicitly mutate headers in options (normalized will keep referencing to the old one)